### DISCUSSION TITLE
How the Predictive Parser separates character strings?

### DISCUSSION THREAD

**Discussion by jw38176** [LangProc/langproc-2021-lab#29] [Q&A]
For a predictive parser, I drew how I understand the process of predicting and parsing: (with its parsing table)

![image](https://user-images.githubusercontent.com/93932554/151172167-2054b5e4-366b-46ff-9d92-37a87764e2a5.png)

![image](https://user-images.githubusercontent.com/93932554/151171378-0ce1c92c-dde6-41f7-9522-b74dd63c157d.png)

Process: the parser would see the first charater, refer to the parsing table(M) with the coordinates M(interpreted terminal, character encountered), select a production rule, and segment the character string accordingly. 

The box in red shows what I'm a bit confused about. For example for **2+5+8$**, the parser sees a number '2' in the beginning and refers to the parsing table. Since this is a top-down parser, it starts with terminal E. Therefore, it refers to element M(E,N) which is E->TE', so the parser breaks the characters up to map to the 2 terminals T and E'. 

The confusion I have is how the parser breaks them up. Under my current logic the parser breaks **2+5+8$** into **2$** and **+5+8$**, i.e. inserting a '$' sign after the '2'. The reasoning for this is that if we try to parse a singular '2' downwards, the parser would refer to element M(T,N) which gives T->FT'. If the '2' maps to the F, then what maps to the T'? 

My interpretation is that there is a '$' inserted in the end so that the parser maps it to T', and when the parser processes refering to M(T',$) it uses T'->ε. (and the '+' is an exception since it is declared in the production rule) 

Having the parser insert the '$' seems very counter-intuitive, but I don't know in what other way the parser solves this problem... Does anyone have any ideas? 

Looking forward to hearing what you guys think! :))

**Comment by johnwickerson** [comment]
Hi, thanks very much for the question. The maiden entry on the Discussions board -- how exciting!

Here's my attempt at a "log" that records the "thoughts" of the predictive parser as it works through (a slightly simpler version of) your example. Hopefully it's helpful? I can't claim to have understand every aspect of your working, but perhaps one difference between your version and mine is that I'm grouping the non-terminals together like `F T' E'`, rather than building a tree where they are separated.

```
Input: 2+5$

Trying to parse E $ and seeing an N, so "E -> T E'".

Now trying to parse T E' $ and seeing an N, so "T -> F T'".

Now trying to parse F T' E' $ and seeing an N, so "F -> N".

Now trying to parse N T' E' $ and seeing an N, so consume it.

Now trying to parse T' E' $ and seeing a +, so "T' -> e".

Now trying to parse E' $ and seeing a +, so "E' -> + T E'".

Now trying to parse + T E' $ and seeing a +, so consume it.

Now trying to parse T E' $ and seeing an N, so "T -> F T'".

Now trying to parse F T' E' $ and seeing an N, so "F -> N".

Now trying to parse N T' E' $ and seeing an N, so consume it.

Now trying to parse T' E' $ and seeing $, so "T' -> e".

Now trying to parse E' $ and seeing $, so "E' -> e".

Now trying to parse $ and seeing $, so we're done.
```
